### PR TITLE
Crangler

### PR DESCRIPTION
**Pull request by kroening**
This adds a new command-line tool, which replaces current flows that enact changes to the program that is analysed. The new flow is to make these changes before parsing, which enables more far-reaching changes and changes that depend on scoped identifiers.

- [X] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Comment by tautschnig** [comment]
Two high-level comments:
1) Do we really need YAML? Couldn't we use JSON instead and have people use tools like `yq` to convert?
2) Even if not a particularly elaborate regression test, could we at least have an example of an application?

File: src/crangler/crangler_parse_options.cpp [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+#include <iostream>
+
+#include "c_wrangler.h"
+
+int crangler_parse_optionst::doit()
+{
+  if(cmdline.args.empty())
+  {
+    std::cerr << "please give a configuration file\n";
```
Comment (tautschnig) [inline src/crangler/crangler_parse_options.cpp:28]: Use error logging via the message handler instead?

File: src/crangler/Makefile [hunk 2]
```diff
@@ -0,0 +1,36 @@
+SRC = c_defines.cpp \
```
Comment (tautschnig) [inline src/crangler/Makefile:1]: Somewhat arbitrary place for comment: we're missing a CMake configuration for this new tool.

File: src/crangler/c_defines.cpp [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    if(!has_prefix(line, "#define "))
+      continue;
+
+    auto space_pos = line.find(' ', 8);
+    if(space_pos == std::string::npos)
+      continue;
+
+    auto id = line.substr(8, space_pos - 8);
+    auto value = line.substr(space_pos + 1, std::string::npos);
+    map[id].value = value;
```
Comment (tautschnig) [inline src/crangler/c_defines.cpp:37]: Do we need to care about `#undef` and a possible re-definition?

File: src/crangler/c_defines.cpp [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+      if(m_it != map.end())
+      {
+        out << m_it->second.value;
+      }
+      else
+        out << t.text;
+    }
+    else
+      out << t.text;
+  }
```
Comment (tautschnig) [inline src/crangler/c_defines.cpp:64]: What about macros that use other macros? What about macros that take parameters (see `getc_unlocked(fp)` example in the comment)?

File: src/crangler/c_defines.h [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+class c_definest
+{
+public:
+  struct definet
+  {
+    optionalt<std::vector<std::string>> parameters;
```
Comment (tautschnig) [inline src/crangler/c_defines.h:28]: Never used (although they'd need to be taken into account).

File: src/crangler/c_defines.h [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+#ifndef CPROVER_CRANGLER_C_DEFINES_H
+#define CPROVER_CRANGLER_C_DEFINES_H
+
+#include <util/optional.h>
+
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+class c_definest
```
Comment (tautschnig) [inline src/crangler/c_defines.h:23]: A brief documentation of the purpose of this class would be useful.

File: src/crangler/c_wrangler.cpp [hunk 7]
```diff
@@ -0,0 +1,474 @@
...
+
+  void configure_sources(const jsont &);
+  void configure_functions(const jsont &);
+  void configure_output(const jsont &);
+};
+
+void c_wranglert::configure_sources(const jsont &config)
+{
+  // look for the 'sources' key
+  auto sources = config["sources"];
```
Comment (tautschnig) [inline src/crangler/c_wrangler.cpp]: There aren't a lot of comments in this code, but then this one stating the obvious. Perhaps I'm missing something and it's not actually doing the obvious?

File: src/crangler/c_wrangler.cpp [hunk 8]
```diff
@@ -0,0 +1,474 @@
...
+    for(const auto &source : to_json_array(sources))
+    {
+      if(!source.is_string())
+        throw deserialization_exceptiont("source must be string");
+
+      this->source_files.push_back(source.value);
+    }
+  }
+
+  // look for the 'includes' key
```
Comment (tautschnig) [inline src/crangler/c_wrangler.cpp]: See above.

File: src/crangler/c_wrangler.cpp [hunk 9]
```diff
@@ -0,0 +1,474 @@
...
+      else if(split[0] == "stub")
+      {
+        std::string rest;
+        for(std::size_t i = 1; i < split.size(); i++)
+        {
+          if(i != 1)
+            rest += ' ';
+          rest += split[i];
+        }
+
```
Comment (tautschnig) [inline src/crangler/c_wrangler.cpp]: Use `join_strings`?

File: src/crangler/c_wrangler.cpp [hunk 10]
```diff
@@ -0,0 +1,474 @@
...
+        (split[0] == "for" && split.size() >= 3 && split[2] == "invariant") ||
+        (split[0] == "while" && split.size() >= 3 && split[2] == "invariant"))
+      {
+        std::string rest;
+        for(std::size_t i = 3; i < split.size(); i++)
+        {
+          if(i != 3)
+            rest += ' ';
+          rest += split[i];
+        }
```
Comment (tautschnig) [inline src/crangler/c_wrangler.cpp]: Use `join_strings`?

File: src/crangler/c_wrangler.cpp [hunk 11]
```diff
@@ -0,0 +1,474 @@
...
+      }
+      else if(split[0] == "assert" && split.size() >= 3)
+      {
+        std::string rest;
+        for(std::size_t i = 2; i < split.size(); i++)
+        {
+          if(i != 2)
+            rest += ' ';
+          rest += split[i];
+        }
```
Comment (tautschnig) [inline src/crangler/c_wrangler.cpp]: Use `join_strings`?

File: src/crangler/c_wrangler.cpp [hunk 12]
```diff
@@ -0,0 +1,474 @@
...
+        split[0] == "ensures" || split[0] == "requires" ||
+        split[0] == "assigns")
+      {
+        std::string rest;
+        for(std::size_t i = 1; i < split.size(); i++)
+        {
+          if(i != 1)
+            rest += ' ';
+          rest += split[i];
+        }
```
Comment (tautschnig) [inline src/crangler/c_wrangler.cpp]: Use `join_strings`?

File: src/crangler/c_wrangler.cpp [hunk 13]
```diff
@@ -0,0 +1,474 @@
...
+
+void c_wranglert::configure_output(const jsont &config)
+{
+  auto output = config["output"];
+
+  if(output.is_null())
+    return;
+
+  if(!output.is_string())
+    throw deserialization_exceptiont("outputs entry must be string");
```
Comment (tautschnig) [inline src/crangler/c_wrangler.cpp]: Perhaps "output" instead of "outputs?"

File: src/crangler/c_wrangler.cpp [hunk 14]
```diff
@@ -0,0 +1,474 @@
...
+  for(const auto &define : c_wrangler.defines)
+    argv.push_back(std::string("-D") + define);
+
+  std::ostringstream result;
+
+  auto run_result = run("cc", argv, "", result, "");
+  if(run_result != 0)
+    throw system_exceptiont("preprocessing " + source_file + " has failed");
+
+  return result.str();
```
Comment (tautschnig) [inline src/crangler/c_wrangler.cpp:340]: Should we have a Windows/MSVC version thereof?

File: src/crangler/c_wrangler.cpp [hunk 15]
```diff
@@ -0,0 +1,474 @@
...
+
+  std::ostringstream result;
+
+  auto run_result = run("cc", argv, "", result, "");
+  if(run_result != 0)
+    throw system_exceptiont("preprocessing " + source_file + " has failed");
+
+  c_definest defines;
+  defines.parse(result.str());
+  return defines;
```
Comment (tautschnig) [inline src/crangler/c_wrangler.cpp:362]: Windows/MSVC?

File: src/crangler/c_wrangler.h [hunk 16]
```diff
@@ -0,0 +1,19 @@
...
+
+\*******************************************************************/
+
+/// \file
+/// C Wrangler
+
+#ifndef CPROVER_CRANGLER_C_WRANGLER_H
+#define CPROVER_CRANGLER_C_WRANGLER_H
+
+#include <util/json.h>
```
Comment (tautschnig) [inline src/crangler/c_wrangler.h]: `class jsont;` might do?

File: src/crangler/mini_c_parser.cpp [hunk 17]
```diff
@@ -0,0 +1,373 @@
...
+  if(tokens.empty())
+    return {};
+
+  DATA_INVARIANT(is_eof(tokens.back()), "token stream must end on eof");
+
+  c_translation_unitt result;
+
+  while(!eof())
+    result.push_back(parse_declaration());
+  // std::cout << "TOKEN: " << consume_token() << "\n";
```
Comment (tautschnig) [inline src/crangler/mini_c_parser.cpp]: Remove?

**Review by tautschnig — APPROVED** [review]
Approving after some further out-of-band discussion. In addition to my earlier comments: some further tests and/or documentation would be really helpful.

File: src/crangler/c_defines.h (hunk 6 above)
Comment (kroening) [reply src/crangler/c_defines.h:23]: added a comment

File: src/crangler/c_wrangler.cpp (hunk 9 above)
Comment (kroening) [reply src/crangler/c_wrangler.cpp]: done

File: src/crangler/c_wrangler.cpp (hunk 10 above)
Comment (kroening) [reply src/crangler/c_wrangler.cpp]: done

File: src/crangler/c_wrangler.cpp (hunk 11 above)
Comment (kroening) [reply src/crangler/c_wrangler.cpp]: done

File: src/crangler/c_wrangler.cpp (hunk 12 above)
Comment (kroening) [reply src/crangler/c_wrangler.cpp]: done

File: src/crangler/c_wrangler.cpp (hunk 13 above)
Comment (kroening) [reply src/crangler/c_wrangler.cpp]: done

File: src/crangler/c_wrangler.h (hunk 16 above)
Comment (kroening) [reply src/crangler/c_wrangler.h]: done

File: src/crangler/mini_c_parser.cpp (hunk 17 above)
Comment (kroening) [reply src/crangler/mini_c_parser.cpp]: done

**Comment by tautschnig** [comment]
We might still want to add a capability to perform name mangling, but once done we should be able to get rid of the `export-file-local-symbols` cross-cutting changes.

**Review by TGWDB — APPROVED** [review]
This is an approval for the Cmake/make aspects as these were blocked by OpenSource/Codeowner

**Pull request merged**